proposals/21-11-transformer-annotations.md

```diff
@@ -0,0 +1,253 @@
+- monopole
+- KnVerey
+- yuwenma
+
+**Status**: implementable
+
+## Summary
+
+This proposal is to extend the `buildMetadata.originAnnotations` option in the kustomization to annotate
+generated resources with information about the source of their generators, and to add a new option 
+`buildMetadata.transformerAnnotations` that will add annotations to each resource with information about
+all the transformers that have processed it. These options will help users understand how their output was
+created and debug it accordingly. 
+
+## Motivation
+
+When a user is managing a large number of resources with kustomize, the output of `kustomize build` will be a 
+long stream of yaml resources, and it can be extremely tedious for the user to understand how each resource 
+was created and processed. However, we can provide the user with an option to annotate each resource with 
+origin and transformation data. This information will help kustomize users
+debug their large kustomize configurations. Because these annotations can be used for debugging purposes, we 
+would like them to be verbose and communicate as much information as possible to the user.
+
+Kustomize previously had a [feature request](https://github.com/kubernetes-sigs/kustomize/issues/3979) to add an 
+option to kustomize build to retain the origin of a resource in an annotation, and a final design was settled 
+on after some discussion. It was implemented by [this pull request](https://github.com/kubernetes-sigs/kustomize/pull/4065) 
+and released in kustomize v4.3.0. This feature added a new field to the kustomization file, `buildMetadata`, an 
+array of boolean options such as `originAnnotations`. If `originAnnotations` is set, `kustomize build` will add annotations 
+to each resource describing the resource's origin. 
+
+While the above feature gives us valuable information about where many resources originated from, there are two 
+pieces missing:
+
+- Origin annotations for generated resources. A comment in the original issue proposed an annotation for generated 
+resources. Though this was accepted as part of the proposal, it was never implemented. However, the original proposal 
+for generated resources doesn't account for resources generated from remote bases.
+
+- Everything in kustomize is a transformer, so we would like to have annotations about which transformers, builtin or 
+custom, touched each resource. Builtin transformers include those that are invoked through various kustomization 
+fields. Some of these transformers are: 
+  - AnnotationsTransformer
+  - HashTransformer
+  - ImageTagTransformer
+  - LabelTransformer
+  - LegacyOrderTransformer
+  - NamespaceTransformer
+  - PatchJson6902Transformer
+  - PatchStrategicMergeTransformer
+  - PatchTransformer
+  - PrefixSuffixTransformer
+  - ReplacementTransformer
+  - ReplicaCountTransformer 
+  - ValueAddTransformer
+  
+  Custom transformers are invoked via the `transformers` field in the kustomization file.
```

Comment:
I think deducing what happened is the best solution. We could check the diff of the resource before and after each item in `transformers`. If there's a difference, we add it to the annotation. We can also use this to deduce if a resource has been generated via the `transformers` field, and thus add the origin annotation for these generated resources. 

I'm inclined to do the same for `validators` and `generators` since I assume (though please correct me if I'm mistaken) that it is possible to put a transformer in either of these fields. WDYT?

Reply:
> Trying to think of the simplest possible thing we could do: assume the first thing generated an all others transformed. I.e. if there's no origin annotation yet, add one. If there is, add to the transformers annotation instead. An extension transformer could mess with us by removing the origin annotation and thereby claiming to have generated the thing to begin with... but maybe that's a feature not a bug? i.e. that means the extension is saying "I delete this object and recreated it from scratch".

SGTM. I can include this implementation proposal in the design details. The only potential problem I could see is what if there is a transformer that is essentially a no-op? Do we still want to include it in the annotation?

> I'm not 100% sure and would want to confirm experimentally, but I think at least some generator types do not receive the existing stream as input, so could not possibly transform it. Theoretically we could be discarding the items results from the validators field, but I'd have to check.

Did a quick check and you're right; generators do not receive input and can't transform it, and for validators a copy is made so that validators cannot transform the input either.

Reply:
> The only potential problem I could see is what if there is a transformer that is essentially a no-op? Do we still want to include it in the annotation?

Good question. I'm leaning towards yes.